File: Data Structures & Algorithms/binary-search/submission-1.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        j = len(nums)
        i = 0
        k=(i+j)//2
        if nums[i] == target: return i
        while i!=k:
            n = nums[k]
            if n == target: return k
            if target>n: 
                i = k
            else: j = k
            k=(i+j)//2
        return -1

File: Data Structures & Algorithms/combination-target-sum-ii/submission-1.cpp
class Solution {
public:
    vector<vector<int>> out;
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        sort(candidates.begin(), candidates.end());
        dfs(candidates, target, 0, {});
        return out;
    }
private:
    void dfs(vector<int>& nums, int t, int ind, vector<int> curr) {
        int sum = 0;
        for (int n:curr) sum += n;
        if (sum==t) {
            out.push_back(curr);
            return;
        }
        if (sum>t || ind >= nums.size()) return;
        vector<int> newV = curr;
        newV.push_back(nums[ind]);
        dfs(nums, t, ind+1, newV);
        int i = ind;
        while (nums[i] == nums[ind]) i++;
        dfs(nums, t, i, curr);
        return;
    }
};
